rust/datafusion/src/catalog/information_schema.rs

```diff
@@ -0,0 +1,225 @@
+/// schema that can introspect on tables in the catalog_list
+struct CatalogWithInformationSchema {
+    catalog_list: Arc<dyn CatalogList>,
+    /// wrapped provider
+    inner: Arc<dyn CatalogProvider>,
+}
+
+impl CatalogProvider for CatalogWithInformationSchema {
+    fn as_any(&self) -> &dyn any::Any {
+        self
+    }
+
+    fn schema_names(&self) -> Vec<String> {
+        self.inner
+            .schema_names()
+            .into_iter()
+            .chain(std::iter::once(INFORMATION_SCHEMA.to_string()))
+            .collect::<Vec<String>>()
+    }
+
+    fn schema(&self, name: &str) -> Option<Arc<dyn SchemaProvider>> {
+        if name.eq_ignore_ascii_case(INFORMATION_SCHEMA) {
+            Some(Arc::new(InformationSchemaProvider {
+                catalog_list: self.catalog_list.clone(),
+            }))
+        } else {
+            self.inner.schema(name)
+        }
+    }
+}
+
+/// Implements the `information_schema` virtual schema and tables
+///
+/// The underlying tables in the `information_schema` are created on
+/// demand. This means that if more tables are added to the underlying
+/// providers, they will appear the next time the `information_schema`
+/// table is queried.
+struct InformationSchemaProvider {
+    catalog_list: Arc<dyn CatalogList>,
+}
+
+impl SchemaProvider for InformationSchemaProvider {
+    fn as_any(&self) -> &(dyn any::Any + 'static) {
+        self
+    }
+
+    fn table_names(&self) -> Vec<String> {
+        vec![TABLES.to_string()]
+    }
+
+    fn table(&self, name: &str) -> Option<Arc<dyn TableProvider>> {
+        if name.eq_ignore_ascii_case("tables") {
+            // create a mem table with the names of tables
+            let mut builder = InformationSchemaTablesBuilder::new();
+
+            for catalog_name in self.catalog_list.catalog_names() {
+                let catalog = self.catalog_list.catalog(&catalog_name).unwrap();
+
+                for schema_name in catalog.schema_names() {
+                    let schema = catalog.schema(&schema_name).unwrap();
+                    for table_name in schema.table_names() {
+                        builder.add_base_table(&catalog_name, &schema_name, table_name)
+                    }
+                }
+
+                // Add a final list for the information schema tables themselves
+                builder.add_system_table(&catalog_name, INFORMATION_SCHEMA, TABLES);
+            }
+
+            let mem_table = builder.build();
+
+            Some(Arc::new(mem_table))
+        } else {
+            None
+        }
+    }
+}
+
+/// Builds the `information_schema.TABLE` table row by row
+
+struct InformationSchemaTablesBuilder {
+    catalog_names: StringBuilder,
+    schema_names: StringBuilder,
+    table_names: StringBuilder,
+    table_types: StringBuilder,
+}
+
+impl InformationSchemaTablesBuilder {
+    fn new() -> Self {
+        Self {
+            catalog_names: StringBuilder::new(10),
```

Comment:
why 10?

Reply:
The string builder requires a number here but the number or rows isn't known. I'll add a comment explaining 10 is some arbitrary number

Reply:
Why not use `0` to keep it simple?

Reply:
I guess I felt that `10` was a siimple as `0` -- lol though I don't really care